### ISSUE TITLE
Brian 2.2: failing tests

### ISSUE THREAD

**Issue by sanjayankur31** [brian-team/brian2#1030]
On a fresh install in a python 3.6 virtual environment, some tests fail. To ensure that the installed copy is used, I did the following:

```
$ python3 -m venv ~/brian-virt
$ source ~/brian-virt/bin/activate
$ pip install cython numpy scipy nose setuptools pyparsing jinja2 py-cpuinfo ipython
$ tar -xvf Brian2-2.2.tar.gz 
$ cd Brian2-2.2
$ rm -rf Brian2.egg-info/ brian2/synapses/cythonspikequeue.cpp
$ python setup.py build
$ python setup.py install
```

The installation went off without any errors. Then, to test the installation, I did:
```
$ cd  # get out of the directory
$ ipython3
Python 3.6.6 (default, Jul 19 2018, 14:25:17)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.1.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import brian2

In [2]: brian2.test()
Running tests in /home/asinha/brian-virt/lib64/python3.6/site-packages/Brian2-2.2-py3.6-linux-x86_64.egg/brian2 for targets numpy, cython (excluding long tests)
Testing codegen-independent code
Testing with multiple processes for numpy, cython, codegen_independent
Resetting to default preferences

Running tests that do not use code generation
In [3]: brian2.test()
Running tests in /home/asinha/brian-virt/lib64/python3.6/site-packages/Brian2-2.2-py3.6-linux-x86_64.egg/brian2 for targets numpy, cython (excluding long tests)
Testing codegen-independent code
Testing with multiple processes for numpy, cython, codegen_independent
Resetting to default preferences

Running tests that do not use code generation
E.....WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
.....S...S..............................WARNING    The expression "i1 / 1" divides two integer values. In previous versions of Brian, this would have used either an integer ("flooring") or a floating point division, depending on the Python version and the code generation target. In the current version, it always uses a floating point division. Explicitly ask for an  integer division ("//"), or turn one of the operands into a floating point value (e.g. replace "1/2" by "1.0/2") to no longer receive this warning. [brian2.parsing.bast.floating_point_division]
.................................................................................................................................WARNING    The expression "(i + 1) / N" divides two integer values. In previous versions of Brian, this would have used either an integer ("flooring") or a floating point division, depending on the Python version and the code generation target. In the current version, it always uses a floating point division. Explicitly ask for an  integer division ("//"), or turn one of the operands into a floating point value (e.g. replace "1/2" by "1.0/2") to no longer receive this warning. [brian2.parsing.bast.floating_point_division]
.....................................................S.S.........................................WARNING    The 'independent' state updater is deprecated and might be removed in future versions of Brian. [brian2.stateupdaters.exact.deprecated_independent]
..................................................................................
======================================================================
ERROR: Failure: TypeError ('NoneType' object is not iterable)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/asinha/brian-virt/lib64/python3.6/site-packages/nose/failure.py", line 39, in runTest
    raise self.exc_val.with_traceback(self.tb)
  File "/home/asinha/brian-virt/lib64/python3.6/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/asinha/brian-virt/lib64/python3.6/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/asinha/brian-virt/lib64/python3.6/site-packages/nose/importer.py", line 86, in importFromDir
    if (self.sameModule(old, filename)
  File "/home/asinha/brian-virt/lib64/python3.6/site-packages/nose/importer.py", line 118, in sameModule
    for path in mod.__path__:
TypeError: 'NoneType' object is not iterable

----------------------------------------------------------------------
Ran 352 tests in 57.626s

FAILED (SKIP=4, errors=1)
Running tests for target numpy:
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
SWARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
SSSSS.SSSSSSSSS.SSSSS.S..S.S.S................................................................S................................WARNING    The 'independent' state updater is deprecated and might be removed in future versions of Brian. [brian2.stateupdaters.exact.deprecated_independent]
.........................................................................................................
----------------------------------------------------------------------
Ran 232 tests in 203.595s

OK (SKIP=25)
Running tests for target cython:
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
SWARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
...WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
SS....S.SSSWARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
SS......S.....WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
..WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
....WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
.WARNING    Integrating equations with GSL is still considered experimental [brian2.stateupdaters.GSL]
...................................S........SS.....FFF.......................................................WARNING    The 'independent' state updater is deprecated and might be removed in future versions of Brian. [brian2.stateupdaters.exact.deprecated_independent]
.................................F...............S......................................
======================================================================
FAIL: brian2.tests.test_neurongroup.test_semantics_floor_division
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/asinha/brian-virt/lib64/python3.6/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/asinha/brian-virt/lib64/python3.6/site-packages/Brian2-2.2-py3.6-linux-x86_64.egg/brian2/tests/test_neurongroup.py", line 1563, in test_semantics_floor_division
    assert len(l) == 0
AssertionError

======================================================================
FAIL: brian2.tests.test_neurongroup.test_semantics_floating_point_division
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/asinha/brian-virt/lib64/python3.6/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/asinha/brian-virt/lib64/python3.6/site-packages/Brian2-2.2-py3.6-linux-x86_64.egg/brian2/tests/test_neurongroup.py", line 1598, in test_semantics_floating_point_division
    'ivalue / 3' in l[0][2]))
AssertionError

======================================================================
FAIL: brian2.tests.test_neurongroup.test_semantics_mod
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/asinha/brian-virt/lib64/python3.6/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/asinha/brian-virt/lib64/python3.6/site-packages/Brian2-2.2-py3.6-linux-x86_64.egg/brian2/tests/test_neurongroup.py", line 1629, in test_semantics_mod
    assert len(l) == 0
AssertionError

======================================================================
FAIL: brian2.tests.test_synapses.test_transmission_scalar_delay_different_clocks
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/asinha/brian-virt/lib64/python3.6/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/asinha/brian-virt/lib64/python3.6/site-packages/Brian2-2.2-py3.6-linux-x86_64.egg/brian2/tests/test_synapses.py", line 1046, in test_transmission_scalar_delay_different_clocks
    assert len(l) == 1, 'expected a warning, got %d' % len(l)
AssertionError: expected a warning, got 2

----------------------------------------------------------------------
Ran 232 tests in 223.856s

FAILED (SKIP=14, failures=4)
ERROR: 2/3 test suite(s) did not complete successfully (see above).
```

Output of pip freeze
```
backcall==0.1.0
Brian2==2.2
Cython==0.29
decorator==4.3.0
ipython==7.1.1
ipython-genutils==0.2.0
jedi==0.13.1
Jinja2==2.10
MarkupSafe==1.1.0
mpmath==1.0.0
nose==1.3.7
numpy==1.15.4
parso==0.3.1
pexpect==4.6.0
pickleshare==0.7.5
prompt-toolkit==2.0.7
ptyprocess==0.6.0
py-cpuinfo==4.0.0
Pygments==2.2.0
pyparsing==2.3.0
scipy==1.1.0
six==1.11.0
sympy==1.3
traitlets==4.3.2
wcwidth==0.1.7
```

GSL version:
```
gsl-devel-2.4-3.fc27.x86_64
```

Have I mucked something up here, or are these genuine bugs worth looking into? 

Thanks again!

**Comment by mstimberg** [comment]
I'm not quite sure what the first error (`'NoneType' object is not iterable`) is about, this seems to be a problem in nose itself...? 

The other errors are because you get a warning from Cython (see also this [mailing list post](https://groups.google.com/d/msg/briansupport/s7MgGzbzPX0/Hio58f1IBgAJ)) and the failing tests are tests that care about the number/types of warnings. We've fixed this with PR #1021, so if you use latest Brian 2 from master (`pip install https://github.com/brian-team/brian2/archive/master.zip`), this should normally no longer happen. We'll do a minor release soon that will contain this fix.

A workaround with Brian's current version would be to use an older version of Cython, I think. But then, you can also simply ignore these warnings.

**Comment by sanjayankur31** [comment]
> I'm not quite sure what the first error (`'NoneType' object is not iterable`) is about, this seems to be a problem in nose itself...?
> 

Yes, I wasn't sure either. I'll dig in more to see what's going on there.

> The other errors are because you get a warning from Cython (see also this [mailing list post](https://groups.google.com/d/msg/briansupport/s7MgGzbzPX0/Hio58f1IBgAJ)) and the failing tests are tests that care about the number/types of warnings. We've fixed this with PR #1021, so if you use latest Brian 2 from master (`pip install https://github.com/brian-team/brian2/archive/master.zip`), this should normally no longer happen. We'll do a minor release soon that will contain this fix.
> 
> A workaround with Brian's current version would be to use an older version of Cython, I think. But then, you can also simply ignore these warnings.

I can do that in the virtualenv but not when building for Fedora. I'll ignore these for the time being, and push an update when a new minor release is made. [Brian has now been approved for Fedora](https://bugzilla.redhat.com/show_bug.cgi?id=1649127) as part of the [NeuroFedora initiative](https://fedoraproject.org/wiki/SIGs/NeuroFedora), so as soon as I've imported it into the build system, Fedora users should simply be able to install it with `sudo dnf install python3-brian2`.

Closing this ticket. Thanks again for the clarification.
